### ISSUE TITLE
Lambda was unable to configure access to your environment  variables because the KMS key is invalid for CreateGrant

### ISSUE THREAD

**Issue by kylelaverty** [terraform-providers/terraform-provider-aws#4633]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.7
+ provider.aws v1.17.0
```

### Affected Resource(s)

* aws_lambda_function

### Expected Behavior

Creating a Lambda that uses a role should work.

### Actual Behavior

Instead we are getting the following error sometimes:

```
Error creating Lambda function: InvalidParameterValueException: Lambda was unable to configure access to your environment 
variables because the KMS key is invalid for CreateGrant. Please check your KMS key settings. KMS Exception: InvalidArnException 
KMS Message: ARN does not refer to a valid principal:
arn:aws:sts::[account #]:assumed-role/[build]_[build try #]_[role name]/[build]_[build try #]_[lambda name]
```

### Steps to Reproduce

1. `terraform apply`
2. Luck and a lot of lambdas

### Important Factoids

It is most likely because of this section of code: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_lambda_function.go#L367

It appears to be treating the error as a `NonRetryableError` which does not seem to line up with what actually happens.

Trying to rerun the Terraform eventually works, even though everything is recreated each time.

**Comment by dshmelev** [comment]
The problem is also reproduced on provider "aws" (1.23.0)

**Comment by emmekappa** [comment]
Just have had the same issue

**Comment by denishonig** [comment]
Hi!
Is there any information when it is going to be fixed?

Thanks.

**Comment by simonvanderveldt** [comment]
Just ran into this as well, when applying the plan again it does work, so it's probably a dependency/waiting/retrying issues. Can't really reproduce it though, so not sure how to make this report more specific.

**Comment by bflad** [comment]
A potential fix for this has been merged into master and will release with version 1.36.0 of the AWS provider, likely later today. (I say potential fix because as with many eventual consistency issues, they are hard to consistently reproduce. 😄 )

**Comment by bflad** [comment]
This has been released in [version 1.36.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1360-september-13-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
